pyfr/solvers/euler/kernels/entropyfilter.mako

```diff
@@ -0,0 +1,155 @@
+# -*- coding: utf-8 -*-
+<%inherit file='base'/>
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+<%include file='pyfr.solvers.euler.kernels.entropy'/>
+
+<% inf = 1e20%>
+<% rf_tol = 1e-6%>
+
+<%pyfr:macro name='get_minima' params='u, dmin, pmin, emin'>
+    fpdtype_t d, p, e;
+    fpdtype_t ui[${nvars}];
+
+    dmin = ${inf}, pmin = ${inf}, emin = ${inf};
+
+    for (int i = 0; i < ${nupts}; i++) {
+        for (int j = 0; j < ${nvars}; j++) {
+            ui[j] = u[i][j];
+        }
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p); emin = fmin(emin, e);
+    }
+
+    // If enforcing constraints on fpts/qpts, compute minima on fpts/qpts
+    % if con_fpts:
+    for (int i = 0; i < ${nfpts}; i++) {
+        for (int j = 0; j < ${nvars}; j++) {
+            // Interpolate to fpts
+            ui[j] = 0.0;
+            for (int k = 0; k < ${nupts}; k++) {
+                ui[j] += intfpts[i][k]*u[k][j];
+            }
+        }
+
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p); emin = fmin(emin, e);
+    }
+    % endif
+
+    % if con_qpts:
+    for (int i = 0; i < ${nqpts}; i++) {
+        for (int j = 0; j < ${nvars}; j++) {
+            // Interpolate to qpts
+            ui[j] = 0.0;
+            for (int k = 0; k < ${nupts}; k++) {
+                ui[j] += intqpts[i][k]*u[k][j];
+            }
+        }
+
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p); emin = fmin(emin, e);
+    }
+    % endif
+</%pyfr:macro>
+
+<%pyfr:macro name='apply_filter' params='umodes, ffac, uf, zeta'>
+    // Precompute filter factors
+    % for i in range(nupts):
+    ffac[${i}] = exp(-zeta*${ubdegs2[i]});
+    % endfor
+
+    // Compute filtered solution
+    for (int uidx = 0; uidx < ${nupts}; uidx++) {
+        for (int vidx = 0; vidx < ${nvars}; vidx++) {
```

Comment:
Transpose 2nd and 3rd loops

Reply:
That would require a separate loop to zero the array, wouldn't it?

Reply:
Yes, although the benefit is that nvars are packed closer in memory than nupts, so it gives a nicer access pattern.

Reply:
Can also avoid the for loop for nvars (expand explicitly using Mako).